---
title: "Useful visualizations to know"
output: html_notebook
---

## some example single cell RNA-seq data. 
- in the file '../src/pbmc_cell_expression.tsv.gz', we have some single cell expression data. This is semi-real data from data generated from blood samples from 3 patients and contains information from Peripheral blood mononuclear cells (PBMCs).
- read this data in and take  a look at it 
```{r}
data <- read_tsv('../src/pbmc_cell_expression.tsv.gz')
```


## making Upset plots 
- install `ComplexUpset`
```{r}
install.packages('devtools')
devtools::install_github("krassowski/complex-upset")
```
- We're going to make an upset plot using our single cell data. 
- Single cell data is very sparse, meaning that only few genes are detected in each cell. Select 3 genes from the above dataset, and make an upset plot for whether or not a gene is expressed in a cell. 
- anything above a count of 0 is expressed, and anything with 0 counts is not expressed
- remember that complexUpset needs columns corresponding to true/false for each attribute we are trying to compare. 
- hint: remember that logical operators(>, <, !) can be applied directly to data.frames

```{r}
library(ComplexUpset)
iris_tibble <-  iris %>% as_tibble

measurements <- data[,-(1:4)]
metadata <- data[,1:4]
logical_measurements <-  measurements > 0
data_for_plotting <- bind_cols(metadata,logical_measurements %>% as_tibble )

upset(data = data_for_plotting, 
      intersect = colnames(data_for_plotting)[-(1:4)], 
      min_size = 10
      )

```



## making alluvial plots 
- install `ggalluvial`
```{r}
install.packages('ggalluvial')
```
- a common step in single cell data is computationally labelling cells with their associated cell type. The data has the output of two labelling tools, in columns `cell_type` and `cell_type_relabeled`. 
- make an alluvial plot, where the first column is `cell_type` and  the second column is `cell_type_relabeled`, to visualize the in cell label from tool to tool 

```{r}
library(ggalluvial)
titanic_wide <- as_tibble(Titanic) %>% 
  mutate(passenger_ID = 1:nrow(.)) %>% select(passenger_ID,Freq=n, everything())
titanic_wide 
library(ggalluvial)
ggplot(data = titanic_wide,
       aes(axis1 = Class, axis2 = Sex,
           y = Freq)) +
  geom_alluvium(fill = 'blue') +
  geom_stratum() +
  geom_text(stat = "stratum", aes(label = after_stat(stratum))) +
  theme_minimal() 

data <- data %>% mutate(Freq=1)

ggplot(data = data,
       aes(axis1 = cell_type, axis2 = cell_type_relabeled,
           y = Freq)) +
  geom_alluvium(fill = 'blue') +
  geom_stratum() +
  geom_text(stat = "stratum", aes(label = after_stat(stratum))) +
  theme_minimal() 

```


## ridge plots 
- install ggridges 
```{r}
install.packages('ggridges')
```

- reshape the data from wide to long format, suitable for plotting
- select a gene and make a ridge plot for its expression values for the different cell types
- select multiple genes and make a faceted ridge plot for each gene.

```{r}
library(ggridges)
data_long <-  data[,-c(1, 3, 4)] %>% pivot_longer(-cell_type, names_to= 'Gene', values_to = 'exp')
data_for_plotting <- filter(data_long, Gene == 'FTL')

ggplot(data_for_plotting) + 
     geom_density_ridges_gradient(aes(x=exp, y=cell_type, fill = stat(x)))

```

## Rainclouds 
- install `gghalves`

- working with the long dataset from before, filter out any expression values equal to zero. This will leave us with a long data frame of only expressed cells. 
- select one gene and make a raincloud plot for each cell type. 
- select multiple genes and make a faceted raincloud plot
```{r}

```




