---
title: "GGplot part 2 exercises"
output: html_notebook
---


## Using multiple geoms 

- this is a more typical form a violin plot, showing both the violin distribution and box plot quantiles. Use multiple geoms to try and recreate this plot. This was generated using the `ToothGrowth` data set. Remember you can use the `?` command to look at the parameters for each geom. use the `ggsave` function to save the plot to disk, and set the ouptut dimensions accordingly to match the example

```{r}
knitr::include_graphics('../src/example_vio_box.png')
```


```{r ex_1}
data("ToothGrowth") 
ToothGrowth

```

## faceting plots 

- Remember the bubble plot we made yesterday using the gapminder data set? We're going to re plot it
- This time use `facet_wrap` to split the data up by continent and make a facted plot. Additionally, add labels for the country names for three countries from each continent
    - you will make a separate data.frame for the labels 

```{r ex_2}
library(tidyverse)
library(ggrepel)
install.packages('ggrepel')

data <- read_tsv('../src/gapminder_2007.tsv.gz') 



```


## creating a complicated plot

- Imagine the following scenario - You're assaying the activity of an enzyme and a hyperactive mutant. Your using a reporter assay that measures the activity in "scientific units"(SU).
- You assay the activity over 24 hours, but realize that when you left the lab the AC turned on and lowered the temperature of the room which may have lowered the temperature.
- this data is stored in '../src/Reporter_assay_results.tsv.gz'
- Use this data to recreate the following plot

```{r}
knitr::include_graphics('../src/reporter_assay_plot.png')
```

This is a bit a of a tough one,so some hints: 
- you will need to reshape the data from wide to long, and calculate the mean, standard deviation, and number of observations at each time point-condition grouping. (try using groupby and summarise)
    - the `n()` function gives the number of observations in a group
- use the  summarized data to calculate standard error(see slides for formula )
- the following geoms are required: `geom_line`, `geom_point`, `geom_label_repel`, `geom_vline` and `geom_errorbar`. You will have to change several theme settings, and scale at least one aesthetic for one of the geoms

- to change the point shapes, you will need to use `scale_shape_manual`. We have not directly covered how to do this, but for some real-life practice, try googling it.

```{r ex_3}
data <- read_tsv('../src/Reporter_assay_results.tsv.gz')

```


## dumbell plots
- this is dumbbell plot. 
![](../src/dumbell_plot.png)

```{r}
knitr::include_graphics('../src/dumbell_plot.png')
```

- the are analogous to a dodged bar plot, where the height, or in this case width, corresponds to the number of cases 
- the data for this plot is in the file '../src/health.csv.gz'
couple hints: 
    - you will need to sort this data before plotting it. Remember how discrete axes are ordered in ggplot with factors.
    - think about what geoms you will have to use( there is no `geom_dumbell`); it will use two geoms we have previously covered. 
    - We have not directly covered how to change axis ticks, but look through the documentation of `scale_x_contionus` for some help
    - the gridlines for the plot can be modified using the `theme` function. Read through the documentation of theme to see how 


```{r ex_4}
library(tidyverse)



```


## Dot plot 

- Dot plots can refer to a variety of plots, but the one we will be covering looks like this 
```{r}
knitr::include_graphics('../src/sc_dot_plot.jpg')
```

- This style of dot plot has four levels of information - two categorical variable(axes) and two continuous variables(color + point size). 
- This particular visualization is generated using single-cell RNA-seq data. The X-axis is an expressed gene, the Y axis is a cell identity(cell type), size is percent expressing, or fraction of cells in each group expressing the gene above a certain threshold, and color is the level of gene expression.


We are going to try and re-create this plot. Here are the general steps to transform the data :

- read the data into R. The file path is '../src/pbmc_cell_expression.tsv.gz'
- reshape the data from wide to long 
- group the data per cell_type per gene_name and summarise to get:
     - Average expression
     - Whether the gene is expressed( is_expressed  = expression > 0)
     - the number of cells in each group ( remember the  `n()` function gives you the size of groups)
- add a column for percent_expressed ( percent_expressed = sum (is_expressed)/ group_size; recall that we can `sum` a logical to get number of `T` occurences)
- From here the data is ready to plot. There is no `geom_dotplot`; you will have to use one of the geoms we have previously covered. 


```{r ex_5}

```






