---
title: "Useful visualizations to know"
output: html_notebook
---

## some example single cell RNA-seq data. 

- in the file '../src/pbmc_cell_expression.tsv.gz', we have some single cell expression data. This is semi-real data from data generated from blood samples from 3 patients and contains information from Peripheral blood mononuclear cells (PBMCs).
- read this data in and take  a look at it 

```{r}
data <- read_tsv('../src/pbmc_cell_expression.tsv.gz')
```

For each of the packages below, I have provided the corresponding documentation and examples website. I recommend you skim through each one and try out an example before trying it out on our data.

## making Upset plots 

- install `ComplexUpset`

```{r}
install.packages('remotes')
remotes::install_github("krassowski/complex-upset")
```

DOCS: https://krassowski.github.io/complex-upset/articles/Examples_R.html

- We're going to make an upset plot using our single cell data. 
- Single cell data is very sparse, meaning that only few genes are detected in each cell. Select 3 genes from the above dataset, and make an upset plot for whether or not a gene is expressed in a cell. 
- anything above a count of 0 is expressed, and anything with 0 counts is not expressed
- remember that `complexUpset` needs columns corresponding to true/false for each attribute we are trying to compare. 
- hint: remember that logical operators(>, <, !) can be applied directly to data.frames

```{r ex_1}
library(ComplexUpset)
data <- read_tsv('../src/pbmc_cell_expression.tsv.gz')
data_m1 <-  data %>% 
    mutate(LYZ_exp  = LYZ > 0, 
           FTL_exp = FTL >0, 
           PF4_exp = PF4 > 0
           )

upset(data = data_m1, intersect= c('LYZ_exp', 'FTL_exp', 'PF4_exp'))
```

```{r}
data_ss <- data[, 7:16]

data_m2 <- data_ss > 0
?upset
upset(as.data.frame(data_m2), intersect = colnames(data_m2), min_size =)
```


## making alluvial plots 
- install `ggalluvial`
```{r ex_2}
install.packages('ggalluvial')
```

DOCS: https://cran.r-project.org/web/packages/ggalluvial/vignettes/ggalluvial.html

- a common step in single cell data is computationally labeling cells with their associated cell type. The data has the output of two labeling tools, in columns `cell_type` and `cell_type_relabeled`. 
- make an alluvial plot, where the first column is `cell_type` and the second column is `cell_type_relabeled`, to visualize the in cell label from tool to tool 

```{r ex_3, fig.height=8}
library(ggalluvial)

data
ggplot(data = data,
       aes(axis1 = cell_type, axis2 = cell_type_relabeled)) +
  geom_alluvium(fill = 'blue') +
  geom_stratum() +
  geom_text(stat = "stratum", aes(label = after_stat(stratum))) +
  theme_minimal() 

```


## ridge plots 
- install ggridges 

```{r}
install.packages('ggridges')
```

DOCS: https://cran.r-project.org/web/packages/ggridges/vignettes/introduction.html

- reshape the data from wide to long format, suitable for plotting
- select a gene and make a ridge plot for its expression values for the different cell types
- select multiple genes and make a faceted ridge plot for each gene.

```{r ex_5}
library(ggridges)

sc_data_summarized <- data %>% 
    select(-barcode, -cell_type_relabeled, -UMAP_1, -UMAP_2, -patient_id) %>% 
    pivot_longer(cols = -cell_type, 
                 names_to = 'gene_name', 
                 values_to ='exp') 
sc_data_1_gene <- sc_data_summarized %>% filter(gene_name == 'FTL')

ggplot(sc_data_summarized, aes(x = exp, y = cell_type, fill = stat(x))) +
  geom_density_ridges_gradient() + 
    facet_wrap(~gene_name)
data("lincoln_weather")


```

## Rainclouds 

- install `gghalves`

```{r}
remotes::install_github('erocoar/gghalves')
```

DOCS: https://erocoar.github.io/gghalves/

- working with the long dataset from before, filter out any expression values equal to zero. This will leave us with a long data frame of only expressed cells. 
- select one gene and make a raincloud (violin + jittered scatterplot)  plot for each cell type. 
- select multiple genes and make a faceted raincloud plot
- Hint: you will need to use the `coord_flip` function to get the plot to be horizontal. use `?coord_flip` to see its usage

```{r ex_6}
library(gghalves)
sc_data_summarized
ggplot(sc_data_summarized, aes(x = cell_type, y = exp, fill = cell_type, color = cell_type)) +
  geom_half_violin(side = 'r') + 
  geom_half_point(range_scale=.5, side = 'l', alpha = .3, shape=16) + 
    coord_flip() + 
    facet_wrap(~gene_name)
```




